content/entry-hash/index.md

```diff
@@ -0,0 +1,219 @@
+   append the result to _hashList_.
+8. Apply the _hashValue_ function to _timestamp_ tagged with `t` (Timestamp). And
+   append the result to _hashList_.
+9. Map over each value _item_ in the _items_ set:
+   1. Apply the _hashValue_ function to _item_ tagged with `r` (Hash).
+10. Sort the elements of _items_, concatenate them and
+11. Apply the _hashValue_ function tagged with `s` (Set). And append the result
+    to _hashList_.
+12. Concatenate the elements of _hashList_ in order (i.e. `[numberHash,
+    keyHash, timestampHash, itemsHash]`), tag it with `l` (List) and hash the result.
+
+
+To walk through the algorithm I'll use the following entry:
+
+```elm
+Entry
+  { number = 6
+  , key = ID "GB"
+  , timestamp = Timestamp 2016 4 5 13 23 5 Utc
+  , item = Set [Sha256 "6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"]
+  }
+```
+
+And this is the current serialisation in JSON:
+
+```json
+[
+  {
+    "index-entry-number":"6",
+    "entry-number":"6",
+    "entry-timestamp":"2016-04-05T13:23:05Z",
+    "key":"GB",
+    "item-hash":["sha-256:6b18693874513ba13da54d61aafa7cad0c8f5573f3431d6f1c04b07ddb27d6bb"]
+  }
+]
+```
+
+Notice they are not exaclty the same even thought they represent the same
+entity. Also notice that the `ID` type is not yet an approved RFC. For this
+RFC this shouldn't be a problem as it works the same as if the type were a
+string.
+
+Each value MUST be tagged by a byte that identifies the type, in a similar way
+as the [objecthash](https://github.com/benlaurie/objecthash) algorithm.
+
+```elm
+type Tag
+  = Hash
+  | Integer
+  | Set
+  | String
+  | Timestamp
+  | List
+
+tagToChar : Tag -> Char
+tagToChar tag =
+  case tag of
+    Hash ->
+      'r'
+    Integer ->
+      'i'
+    Set ->
+      's'
+    String ->
+      'u'
+    Timestamp ->
+      't'
+    List ->
+      'l'
+```
+
+```elm
+hashValue : Alg -> String -> Tag -> Hash
+```
+
+Where:
+
+* `Alg` is the hashing algorithm (e.g. `Sha256`),
+* `String` is the string representation of the value to hash,
+* `Tag` is the tag for the `Value` type and
+* `Hash` is the list of bytes resulting of the hashing algorithm.
+
+Following the JSON example from above, these are the hashes for each value:
+
+```elm
+numberValue = toString 6
+numberValue == "6"
+
+numberHash = hashValue Sha256 numberValue Integer
+toHex numberHash == "396ee89382efc154e95d7875976cce373a797fe93687ca8a27589116644c4bcd"
+```
+
+```elm
+timestampValue = toString (Timestamp 2016 4 5 13 23 5 Utc)
+toHex timestampValue == "2016-04-05T13:23:05Z"
+
+timestampHash = hashValue Sha256 timestampValue Timestamp
+toHex timestampHash == "f22ecc4464f22c8fee624769189665a0afd7ef10a2775a000082c47cbd9f6419"
+```
+
+```elm
+keyValue = toString (ID "GB")
+keyValue == "GB"
+
+keyHash = hashValue Sha256 keyValue String
+toHex keyHash == "fff7021c7df4426be0f9a3c83f236eb6f85d159e624b010d65e6dde267889c21"
+```
+
+The item is a special case:
```

Comment:
Does this need to be a special case? Can't we just treat it as a string and follow the normal objecthash algorithm? If so that would simplify the implementation a lot as you just need to implement the objecthash algorithm once.

Reply:
Well, objecthash accounts for "raw" byte sequences so from that perspective it is not a special case. It is though with the rest of values in Entry. The wording is not clear enough.